kore/src/Kore/Step/AntiLeft.hs

```diff
@@ -0,0 +1,417 @@
+            freeVariables aliasTerm
+            <> fold
+                (  toList (freeVariables <$> maybeInner)
+                ++ map freeVariables leftHands
+                )
+
+instance
+    InternalVariable variable
+    => HasFreeVariables (AntiLeftLhs variable) variable
+  where
+    freeVariables antiLeft@(AntiLeftLhs _ _ _ _) = case antiLeft of
+        AntiLeftLhs { aliasTerm, existentials, predicate, term } ->
+            bindVariables
+                (map (fmap SomeVariableNameElement) existentials)
+                (  freeVariables predicate
+                <> freeVariables term
+                <> freeVariables aliasTerm
+                )
+
+mapVariables
+    :: (Ord variable1, FreshPartialOrd variable2)
+    => AdjSomeVariableName (variable1 -> variable2)
+    -> AntiLeft variable1
+    -> AntiLeft variable2
+mapVariables adj antiLeft@(AntiLeft _ _ _) =
+    case antiLeft of
+        AntiLeft {aliasTerm, maybeInner, leftHands} ->
+            AntiLeft
+                { aliasTerm = TermLike.mapVariables adj aliasTerm
+                , maybeInner = mapVariables adj <$> maybeInner
+                , leftHands = map (mapVariablesLeft adj) leftHands
+                }
+
+mapVariablesLeft
+    :: (Ord variable1, FreshPartialOrd variable2)
+    => AdjSomeVariableName (variable1 -> variable2)
+    -> AntiLeftLhs variable1
+    -> AntiLeftLhs variable2
+mapVariablesLeft adj antiLeft@(AntiLeftLhs _ _ _ _) =
+    case antiLeft of
+        AntiLeftLhs {aliasTerm, existentials, predicate, term} ->
+            AntiLeftLhs
+                { aliasTerm = TermLike.mapVariables adj aliasTerm
+                , existentials = map (mapElementVariable adj) existentials
+                , predicate = Predicate.mapVariables adj predicate
+                , term = TermLike.mapVariables adj term
+                }
+
+substitute
+    :: InternalVariable variable
+    => Map (SomeVariableName variable) (TermLike variable)
+    -> AntiLeft variable
+    -> AntiLeft variable
+substitute subst antiLeft@(AntiLeft _ _ _) =
+    case antiLeft of
+        AntiLeft {aliasTerm, maybeInner, leftHands} ->
+            AntiLeft
+                { aliasTerm = TermLike.substitute subst aliasTerm
+                , maybeInner = substitute subst <$> maybeInner
+                , leftHands = map (substituteLeft subst) leftHands
+                }
+
+substituteLeft
+    :: InternalVariable variable
+    => Map (SomeVariableName variable) (TermLike variable)
+    -> AntiLeftLhs variable
+    -> AntiLeftLhs variable
+substituteLeft subst antiLeft@(AntiLeftLhs _ _ _ _) =
+    case antiLeft of
+        AntiLeftLhs {aliasTerm, existentials, predicate, term} ->
+            AntiLeftLhs
+                { aliasTerm = TermLike.substitute subst' aliasTerm
+                , existentials
+                , predicate = Predicate.substitute subst' predicate
+                , term = TermLike.substitute subst' term
+                }
```

Comment:
Because there can only be one case alternative, we might use `where` instead, but this is not very important and I leave it up to you.

Reply:
That seems better, fixed.